bowtie/_cli.py

```diff
@@ -1671,6 +1695,485 @@ async def info(
+
+    return main_table
+
+
+def _trend_table_in_markdown_for(
+    id: ConnectableId,
+    versions: Set[str],
+    dialects_trend: dict[Dialect, _report.Report],
+) -> str:
+    rows_data: list[list[str]] = []
+    final_content = ""
+
+    inner_table_columns = [
+        "Version",
+        "Skips",
+        "Errors",
+        "Failures",
+    ]
+
+    def create_inner_table_row(
+        version: str,
+        unsuccessful: Unsuccessful,
+    ):
+        return [
+            version,
+            str(len(unsuccessful.skipped)),
+            str(len(unsuccessful.errored)),
+            str(len(unsuccessful.failed)),
+        ]
+
+    for dialect, report in dialects_trend.items():
+        inner_table_rows: list[list[str]] = []
+        if versions:
+            for version, unsuccessful in report.latest_to_oldest():
+                inner_table_rows.append(
+                    create_inner_table_row(version, unsuccessful),
+                )
+        else:
+            implementation = report.implementations[id]
+            version = implementation.version or "latest"
+            unsuccessful = report.unsuccessful(id)
+            inner_table_rows.append(
+                create_inner_table_row(version, unsuccessful),
+            )
+
+        inner_markdown_table = convert_table_to_markdown(
+            inner_table_columns,
+            inner_table_rows,
+        )
+        row_data = [dialect.pretty_name, inner_markdown_table]
+        rows_data.append(row_data)
+
+    final_content += f"## Trend Data of {id} versions:\n\n"
+    for _, row_data in enumerate(rows_data):
+        final_content += f"### Dialect: {row_data[0]}\n\n"
+        final_content += f"{row_data[1]}\n"
+
+    return final_content
+
+
+@subcommand
+@click.option(
+    "--implementation",
+    "-i",
+    "connectable",
+    required=True,
+    multiple=False,
+    default=None,
+    type=_connectables.ClickParam(),
+    metavar="IMPLEMENTATION",
+    help=(
+        "A connectable ID for a JSON Schema implementation supported "
+        "by Bowtie. Run `bowtie filter-implementations` for the full "
+        "list of supported implementations."
+    ),
+)
+@click.option(
+    "--dialect",
+    "-D",
+    "dialects",
+    multiple=True,
+    default=lambda: Dialect.known(),
+    type=_Dialect(),
+    metavar="URI_OR_NAME",
+    help=(
+        "A URI or shortname identifying the dialect of the test suite "
+        "you wish to see the trend data for. Possible shortnames include: "
+        f"{POSSIBLE_DIALECT_SHORTNAMES}. If none specified then defaults "
+        "to all known dialects."
+    ),
+)
+@format_option()
+def trend(
+    connectable: Connectable,
+    dialects: Iterable[Dialect],
+    format: _F,
+):
+    """
+    Show trend data of an implementation across its multiple versions.
+    """
+    id = connectable.to_terse()
+    versions = asyncio.run(download_versions_of(id))
+
+    downloaded_reports = asyncio.run(
```

Comment:
You have a small tendency to sprinkle `asyncio.run` in places -- that's a top level function which starts a loop and runs some main thing -- it's not something that should really appear multiple times in a program -- in other functions we start one via `asyncio.run` before actually running a subcommand, you possibly might be looking for that pattern.

In Javascript this function doesn't really exist (AFAIK), there's a single global event loop -- in Python, there can be many (and not even just one event loop framework).

Reply:
Any suggestions you have for this ? Basically I had to do this way because firstly i need to download the versions for the `ConnectableId` that was passed and then later I want to download reports using that versions metadata that was fetched. So should I do all of this in the `--implementation` option callback itself ? Is that a better place for running `asyncio.run` as you say before actually running a subcommand ?

Reply:
I don't really see the issue, this sounds like normal asynchronous code (and looks like it from the body of this function) -- if you need to do 2 async things, call the async code, await if you need to wait for the result in order to continue, then do the other one, await if need be, etc., just do it all within one event loop rather than starting and stopping an event loop each time you want to do an async thing.